### PR TITLE
[SIDE-461] Convert LDK to TS

### PR DESCRIPTION
**Pull request by richseviora**
This PR:

- Converts the LDK source files to TS.
  - They're now all located under `src`.
  - Adds typings to all params and returns.
  - Replaced `require` statements with `import` statements.
- Outputs JS to the `dist` directory for consumption by consumers, with typings attached for use in TS projects.
- Adds automatic TS definition file generation for the GRPC proto files.
- Adds TS compilation on commit.
- Updates JSDoc to parse TS.
- Add lint-staged as a pre-commit hook to format all js/ts/JSON (via prettier) and check for any linting errors.

Some thoughts:

- ~I'd like to go through and completely type the LDK (parameters, returns, etc) in a separate PR. Not required for immediate consumption by consumers but it'll help further development.~

### CODE REVIEW

**Comment by BillyMurrayOlive** [comment]
@richseviora The linter is throwing a bunch of warning on the following rules:
```
@typescript-eslint/explicit-module-boundary-types
@typescript-eslint/no-explicit-any
```
How do you feel about just disabling those rules?

**Comment by BillyMurrayOlive** [comment]
@richseviora Can you please regenerate the docs?
Since the files have been renamed from `.js` to `.ts`, you will also need to remove the old files.
Just deleting the whole docs folder and regenerating is probably easiest.
```
rm -rf docs && npm run docs
```

**Comment by BillyMurrayOlive** [comment]
I asked three engineers that have worked on some of our other Typescript projects to take a look at this PR.
I asked them to make sure we are following any standards or best practices used in these other projects.
They are planning to review it this afternoon.

**Comment by richseviora** [comment]
Awesome! I'll rebuild the docs + go through the longer typing exercise as discussed.

**Comment by BRKurek** [comment]
Overall, everything looks good to me. The only comment I have is it looks like there’s a heavy use of the `any` type. I think that’s okay for now because this is an initial migration (and you even called out that you would like to go through and completely type everything in a future PR), but most of our TS project that we build from scratch rarely use any (for good reason). Just something to keep an eye on going forward

**Comment by reidgiff** [comment]
I agree, everything looks good. I will call out that we are more strict on linting and generally prefer multiple files for multiple classes, but it looks like it makes sense in context here.

**Comment by richseviora** [comment]
Thanks for the feedback! 😃 

I'm going through the typing exercise right now.

As part of the changes I'll be pushing up I'll be adding lint-staged to start linting files before commit and fix all the implicit types.

**Comment by richseviora** [comment]
Typing exercise complete!

**Comment by richseviora** [comment]
I've kept `any` in a few places where it either made sense (`console` function arguments), or where the grpc library had some expectations I couldn't avoid without changing the code itself.

File: scripts/protoFixes.js [hunk 1]
```diff
@@ -2,30 +2,58 @@ const fs = require('fs');
 const path = require('path');
 
 // load files that need edited
-const brokerGrpcPbPath = path.join(__dirname, '../proto/broker_grpc_pb.js');
-const healthGrpcPbPath = path.join(__dirname, '../proto/health_grpc_pb.js');
-const ldkGrpcPbPath = path.join(__dirname, '../proto/ldk_grpc_pb.js');
-const stdioGrpcPbPath = path.join(__dirname, '../proto/stdio_grpc_pb.js');
+const protoDir = path.join(__dirname, '../src/proto');
```
Comment (BillyMurrayOlive) [inline scripts/protoFixes.js:5]: This was a smart addition.

File: .jsdoc.json [hunk 2]
```diff
@@ -8,5 +8,12 @@
     "default": {
       "includeDate": false
     }
+  },
+  "source": {
+    "includePattern": ".+\\.(js|ts)$"
+  },
+  "plugins": ["node_modules/better-docs/typescript"],
```
Comment (BillyMurrayOlive) [inline .jsdoc.json:15]: This looks nice. Does anything need changed in the `doc` command in `package.json`?

Comment (richseviora) [reply .jsdoc.json:15]: It's building so I think it's fine as-is for now :) Though as mentioned I'm going to take a look at some different options for doc builders that use TS

File: scripts/protoFixes.js (hunk 1 above)
Comment (richseviora) [reply scripts/protoFixes.js:5]: 😄

**Pull request merged**